### PR TITLE
Add code formatting and checking targets

### LINKED ISSUE #38: Linter & Auto code formatting
some kind of linting support and especially auto code formatting would be desirable.
Could be integrated into CI and as commit hook.

### PR DESCRIPTION
**Pull request by ikstream**
Signed-off-by: Stefan Venz <stefan.venz@protonmail.com>

## Related Issues & PRs

fix #38

## Description

- [x] Add separate Docker for code processing
- [x] Make targets added
- [x] Formatting file added

## Context

Add Code formatting and static code analysis to improve overall code quality

## Testing

- [x] tested by me
- [ ] Feedback about code structure after formatting

To test this PR: 
change to the flavor you want to check e.g. `cd lifensensor`
`make check-code` to see static code analysis (nothing wrong with the code so far)
`make format-code` to apply new formatting style

## Documentation

- [x] Documented both new features

### CODE REVIEW

File: .Makefile.mk [hunk 1]
```diff
@@ -31,6 +31,9 @@ TEMPLATEPATH=../.Makefile.template
 ### default target ###
 DEFAULT += help
 
+# header and source files proccessed by format and check
+FILES=$(shell cd ../; find -regex '.*\.\(c\|h\)' -not -path *build*)
```
Comment (nopeslide) [inline .Makefile.mk]: why is this variable not in the variables section?
why determine files here and not inside the container?
why the `cd ..` when using `find`?

File: .Makefile.mk [hunk 2]
```diff
@@ -107,6 +110,32 @@ CHECK += check-monitor
...
+		--std=c11 \
+		--verbose \
+		--suppress=missingInclude \
+		--template="[{severity}][{id}] {message} {callstack} (On {file}:{line})"'
+
+### clang-format
+
+.PHONY: format-code
+TARGET += format-code
+ALL += format-code
```
Comment (nopeslide) [inline .Makefile.mk]:
```suggestion
```
code format shouldn't be any automatic process

File: .Makefile.mk [hunk 3]
```diff
@@ -107,6 +110,32 @@ CHECK += check-monitor
 HELP_check-monitor = check env if monitor is possible
 check-monitor: | check-docker check-dev
 
+### cppcheck  target
+
+.PHONY: check-code
+TARGET += check_code
+ALL += check-code
+HELP_build = checks code with cppcheck
```
Comment (nopeslide) [inline .Makefile.mk]:
```suggestion
HELP_check-code = checks code with cppcheck
```

File: .Makefile.mk [hunk 4]
```diff
@@ -107,6 +110,32 @@ CHECK += check-monitor
...
+		--verbose \
+		--suppress=missingInclude \
+		--template="[{severity}][{id}] {message} {callstack} (On {file}:{line})"'
+
+### clang-format
+
+.PHONY: format-code
+TARGET += format-code
+ALL += format-code
+HELP_build = formats code with clang-format
```
Comment (nopeslide) [inline .Makefile.mk]:
```suggestion
HELP_format-code = formats code with clang-format
```

File: docker/Makefile [hunk 5]
```diff
@@ -76,6 +87,21 @@ vscode-image:  qemu-image | check-docker
 		vscode
 
 ### docker run targets ###
+.PHONY: format
+TARGET += format
+HELP_idf = runs formatting container
```
Comment (nopeslide) [inline docker/Makefile]:
```suggestion
HELP_format = runs formatting container
```

File: docker/Makefile [hunk 6]
```diff
@@ -76,6 +87,21 @@ vscode-image:  qemu-image | check-docker
...
+format: | format-image check-docker
+	docker run \
+		--rm \
+		-ti \
+		$(DOCKER_DEV) \
+		-v format-home:/home/developer:rw \
```
Comment (nopeslide) [inline docker/Makefile:98]: is a persistent home needed?

**Review by nopeslide — CHANGES_REQUESTED** [review]
is  there a way to check if the source code conforms with the format?
if there is, we should integrate such check as target in the makefile

File: docker/Makefile (hunk 6 above)
Comment (ikstream) [reply docker/Makefile:98]: As we want to change our src files, i think so

File: .Makefile.mk (hunk 1 above)
Comment (ikstream) [reply .Makefile.mk]: if i tried to solve it in find, sometimes no files where returned. unfortunately it wasn't a deterministic problem.

Comment (ikstream) [reply .Makefile.mk]: It was initially in the variables sections, seems to be moved during rebase. will fix that

File: docker/Makefile (hunk 6 above)
Comment (nopeslide) [reply docker/Makefile:98]: I don't see anything that is created in the home directory,
therefore I don't see the necessity.
Did I overlook sth?
If not: also remove the clean-format

Comment (ikstream) [reply docker/Makefile:98]: the c and header files are adapted to the style, nothing is created

Comment (nopeslide) [reply docker/Makefile:98]: the files this container modifies are mounted and have nothing to do with the container home dir. the other docker containers feature a persistent home for custom configs and caches of the toolchain (i.e. ccache, .gdbinit, ...). this is not needed in this case (why customize the format style?), hence my initial question.

**Comment by jhonnyam** [comment]
Would be nice to get some feedback which files would be/where changed.

**Comment by nopeslide** [comment]
```diff
-typedef struct {
-       const char *name;
-       StaticTask_t tcb;
-       TaskHandle_t handle;
-       StackType_t *stack;
-       UBaseType_t priority;
+typedef struct
+{
+    const char * name;
+    StaticTask_t tcb;
+    TaskHandle_t handle;
+    StackType_t *stack;
+    UBaseType_t  priority;
 } _spo2_task;
```
```diff
-typedef struct {
-       UBaseType_t length;
-       UBaseType_t item_size;
-       uint8_t *buffer;
-       StaticQueue_t queue;
-       QueueHandle_t handle;
+typedef struct
+{
+    UBaseType_t   length;
+    UBaseType_t   item_size;
+    uint8_t *     buffer;
+    StaticQueue_t queue;
+    QueueHandle_t handle;
 } _spo2_queue;
```
the pointer 'star' placement seems to be kind of random.

**Comment by nopeslide** [comment]
```diff
-       spo2_queue->handle = xQueueCreateStatic(
-               spo2_queue->length,
-               spo2_queue->item_size,
-               spo2_queue->buffer,
-               &spo2_queue->queue
-       );
+    spo2_queue->handle =
+        xQueueCreateStatic(spo2_queue->length, spo2_queue->item_size,
+                           spo2_queue->buffer, &spo2_queue->queue);
```
the current settings seem to stuff as much parameters as possible in one line.
if we break up parameters over multiple lines shouldn't we give each parameter its own line?

**Comment by nopeslide** [comment]
```diff
-static gpio_config_t gpio_agc_inc_config = {
-       .pin_bit_mask = SPO2_DCP_INC_PIN,
-       .mode = GPIO_MODE_OUTPUT,
-       .pull_down_en = GPIO_PULLDOWN_ENABLE,
-       .pull_up_en = GPIO_PULLUP_DISABLE,
-       .intr_type = GPIO_PIN_INTR_DISABLE
-};
+static gpio_config_t gpio_agc_inc_config = {.pin_bit_mask = SPO2_DCP_INC_PIN,
+                                            .mode         = GPIO_MODE_OUTPUT,
+                                            .pull_down_en =
+                                                GPIO_PULLDOWN_ENABLE,
+                                            .pull_up_en = GPIO_PULLUP_DISABLE,
+                                            .intr_type = GPIO_PIN_INTR_DISABLE};
```
multi-line initializations are hard to read when attributes don't get their own line. probably the same issues as with function parameters

**Comment by nopeslide** [comment]
```diff
-       sample->ird_dc = spo2_dc_lowpass(sample->ird_dc,
-                                        (uint16_t)raw_data->ird_dc);
-       sample->red_dc = spo2_dc_lowpass(sample->red_dc,
-                                        (uint16_t)raw_data->red_dc);
+    sample->ird_ac =
+        spo2_ac_lowpass_no_offset(sample->ird_ac, (uint16_t)raw_data->ird_ac);
+    sample->red_ac =
+        spo2_ac_lowpass_no_offset(sample->red_ac, (uint16_t)raw_data->red_ac);
```
breaking up assignments should be a last resort, would prefer parameter breaking here.

**Comment by nopeslide** [comment]
```diff
-Counter
-counter_reset
-(void);
+Counter counter_reset(void);
```
Giving function declarations/definitions at least an extra line for the return value seemed like a good idea

**Comment by nopeslide** [comment]
> Would be nice to get some feedback which files would be/where changed.

this would be possible by moving the find instruction inside the container. will look into it

**Comment by nopeslide** [comment]
> Would be nice to get some feedback which files would be/where changed.

not absolutely happy with the implementation, but here you go
`make check-format` lists all files that need formatting and errors if any file needs formatting
`make format-code` now lists all files it touches (regardless if format was needed)

**Comment by nopeslide** [comment]
> not absolutely happy with the implementation, but here you go
> `make check-format` lists all files that need formatting and errors if any file needs formatting
> `make format-code` now lists all files it touches (regardless if format was needed)

on second thought, I think these targets are unneeded. If you format your code git shows what happened. if you want to check if any code needs formatting, format it and ask git if changes occurred.

**Comment by ikstream** [comment]
We could run `clang-format` twice, once with `-i` and once without. This would show the changes prior applying it and only one command would be needed.

**Comment by nopeslide** [comment]
> We could run `clang-format` twice, once with `-i` and once without. This would show the changes prior applying it and only one command would be needed.

`clang-format` dumps the whole file, making it kind of useless output. would prefer something like this
to check format (line diff)
```bash
git ls-files -ico -x '**/*.c' -x '**/*.h' | xargs -I % sh -c "clang-format % | git diff --exit-code --no-index --color % -"
```
to check format (word diff, fancy for these minimal whitespace changes)
```bash
git ls-files -ico -x '**/*.c' -x '**/*.h' | xargs -I % sh -c "clang-format % | git -c color.diff.new='normal 22' -c color.diff.old='normal 166' diff --exit-code --no-index --color --word-diff=color --word-diff-regex=. % -"
```
image

to apply format
```bash
git ls-files -ico -x '**/*.c' -x '**/*.h' | xargs -I % clang-format % 
```

**Comment by ikstream** [comment]
As there seems to be something broken with the config, besides the unfix-able problem, I will close this in favor of #70

**Pull request closed without merging**